src/datasources/locking.ts

```diff
@@ -0,0 +1,98 @@
+import { LockCreate, Relock, Delegate, Withdraw } from "../../generated/schema";
+
+import {
+  LockCreate as LockCreateEvent,
+  Relock as RelockEvent,
+  Delegate as DelegateEvent,
+  Withdraw as WithdrawEvent,
+} from "../../generated/locking/Locking";
+
+import { events, transactions } from "@amxx/graphprotocol-utils";
+
+import { fetchAccount } from "../fetch/account";
+
+import { fetchLocking, fetchLock } from "../fetch/locking";
+
+export function handleLockCreate(event: LockCreateEvent): void {
+  let locking = fetchLocking(event.address);
+
+  let lock = fetchLock(locking, event.params.id);
+  lock.owner = fetchAccount(event.params.account).id;
+  lock.delegate = fetchAccount(event.params.delegate).id;
+  lock.time = event.params.time;
+  lock.amount = event.params.amount;
+  lock.slope = event.params.slopePeriod.toI32();
+  lock.cliff = event.params.cliff.toI32();
+  lock.relocked = false;
+  lock.save();
+
+  let ev = new LockCreate(events.id(event));
+  ev.emitter = locking.id;
+  ev.transaction = transactions.log(event).id;
+  ev.timestamp = event.block.timestamp;
+  ev.locking = lock.locking;
+  ev.lock = lock.id;
+  ev.owner = event.params.account;
+  ev.save();
+}
+
+export function handleRelock(event: RelockEvent): void {
+  let locking = fetchLocking(event.address);
+
+  let oldLock = fetchLock(locking, event.params.id);
+  let newLock = fetchLock(locking, event.params.counter);
+  oldLock.relocked = true;
+  oldLock.replacedBy = newLock.id;
+  oldLock.save();
+
+  newLock.owner = fetchAccount(event.params.account).id;
+  newLock.delegate = fetchAccount(event.params.delegate).id;
+  newLock.time = event.params.time;
+  newLock.amount = event.params.amount;
+  newLock.slope = event.params.slopePeriod.toI32();
+  newLock.cliff = event.params.cliff.toI32();
+  newLock.replaces = oldLock.id;
+  newLock.relocked = false;
+  newLock.save();
+
+  let ev = new Relock(events.id(event));
+  ev.emitter = locking.id;
+  ev.transaction = transactions.log(event).id;
+  ev.timestamp = event.block.timestamp;
+  ev.locking = newLock.locking;
+  ev.newLock = newLock.id;
+  ev.oldLock = oldLock.id;
+  ev.owner = newLock.owner;
+  ev.save();
+}
+
+export function handleDelegate(event: DelegateEvent): void {
+  let locking = fetchLocking(event.address);
+
+  let lock = fetchLock(locking, event.params.id);
+  lock.delegate = fetchAccount(event.params.delegate).id;
+  lock.save();
+
+  let ev = new Delegate(events.id(event));
+  ev.emitter = locking.id;
+  ev.locking = locking.id;
```

Comment:
is this correct? ditto L92

Reply:
It's debatable wether there would be a better way to structure this, but I've followed the same pattern that the `openzeppelin-subgraph` modules use, because we're extending from that so it's nice to be consistent.
Check here: https://github.com/OpenZeppelin/openzeppelin-subgraphs/blob/main/src/datasources/governor.ts#L167-L176

Basically, the structure keeps a few different entity types:
- Contracts - like "Governor" (for Governor.sol), or "Locking"  for (Locking.sol)
- Domain-specific entities - like a "Lock" or a "Proposal"
- Event entities - which map 1:1 to Contract events
There's also a global "Account" type, which represents any address in the subgraph.

Then how it's structure is that an event has an `emitter` which represents the address that emitted and it's a connection of type account: https://github.com/OpenZeppelin/openzeppelin-subgraphs/blob/main/src/datasources/governor.gql.json#L77-L88
But also an event specific connection which is casted to the actual type, so in our example `locking` of type `Locking`, and you can check the link above for an example with `governor` of type `Governor`.
What this allows us is to have the inverse relationship, which can be seen here for Governor: https://github.com/OpenZeppelin/openzeppelin-subgraphs/blob/main/src/datasources/governor.gql.json#L18-L21 and we have the same thing in this PR. this lets us do something like:

```graphql
locking(id: "<address>") {
  delegate { 
    id
  }
}
```

To query all entity of types `delegate` i.e. all delegate events starting from the locking contract at a specific address.

I think the reason for this duplication of `emitter` and `locking` which are essentially the same thing, is a limitation of type relationships in graphql, I didn't investigate it further but decided to follow the same structure as the OZ ones because otherwise even if I find a better way it might get confusing if the graph structure isn't consistent.

As for the question about difference between `locking` and `lock`, I think that's also clear now. A `lock` is actually the domain-specific entity that the event is related to, i.e. when we delegate we delegate from a specific lock.